### DISCUSSION TITLE
REPL I/O flushing in sly

### DISCUSSION THREAD

**Discussion by mustiboost** [joaotavora/sly#433] [Ideas]
Given the test code: 

```lisp
(dotimes (x 10)
	       (sleep 1) (format t "~a " x))
```
sly REPL waits for 10 seconds and then prints out 

```shell
0 1 2 3 4 5 6 7 8 9
```
all at once, instead of one by one every second as slime, or slima (on atom), or even ielm/elisp (just mentioning elisp because i'm talking from a user perspective and it's a lisp that has common roots with CL). 

On the sbcl repl you start in the command line it's different and all the numbers appear at once after 10 seconds. You can use `(finish-output)` in both sly and sblc command line repl to achieve the desired results, of course, but I think that should be unnecessary. Also most people do not use just sbcl or any other "inferior lisp" on the command line. I think even without a `(finish-output)` call the numbers should show up 1 second apart on the same line. That makes the most sense.  It does not seem like a big deal in this example, but who knows where else this "issue" may crop up, in contexts that are much more critical. I mean, think about it, every time (format t ...) is called something should show up on the screen and not just wait in the repl io buffer, unless we really want it to wait there, right? 

What do you guys think?  Any major drawbacks to my suggestion?

**Comment by joaotavora** [comment]
I think SLIME uses a so-called `auto-flush-thread` that uses its a thread per stream and runs every 0.2 seconds flushing the stream.  I remember taking it out cause I didn't much see the point of it and because it seems as bit wasteful (not to mention confusing newbies in the thread view).

We could add it back but I would hope for some hook provided by the implementation directly.

There is also `SLYNK:*DEDICATED-OUTPUT-STREAM-BUFFERING*`, but I don't know how well it works for all implementations, and it's for sockets, not exactly streams (hence the "dedicated" in the name, it only works for that alternative).

Open to suggestions.  FTR the SBCL REPL doesn't do what @mustiboost wants, but the Allegro REPL does.

**Reply by rprimus** [reply]
Mon Mar 22 13:42:35 GMT 2021

`CCL REPL` and `LispWorks listener` also do the `expected`.   Seems `SBCL` is the anomaly (thus far).
